public/nextjs_migration/client/js/circleChart.js

```diff
@@ -0,0 +1,326 @@
+    this.title = ''
+
+    this.svg = null
+    this.total = 0
+    this.updateTimeout = null
+
+    this.render()
+  }
+
+  /**
+   * @param {string} name
+   * @param {string} oldValue
+   * @param {string} newValue
+   */
+  attributeChangedCallback (name, oldValue, newValue) {
+    if (newValue === 'undefined' || newValue === oldValue) {
+      return
+    }
+
+    switch (name) {
+      case 'data':
+        this.data = JSON.parse(newValue)
+        this.createOrUpdateChart()
+        break
+      case 'show-percent-for':
+        this.showPercentFor = newValue
+        break
+      case 'title':
+        this.title = newValue
+        break
+      default:
+        console.warn(`Unhandled attribute: ${name}`)
+    }
+  }
+
+  composeCircles () {
+    let sliceOffset = 0
+    /** @type {string[]} */
+    const init = []
+    return `
+      ${this.data?.reduce((acc, curr) => {
+        const percentage = calcPercentage(this.total, curr.count)
+        const innerRadius = this.showPercentFor !== '' ? 0.85 : 0
+        const strokeLength = CHART_CIRCUMFERENCE * percentage
+
+        const circle = `
+          <circle
+            stroke='${curr.color}'
+            stroke-dasharray='${strokeLength} ${CHART_CIRCUMFERENCE}'
+            stroke-dashoffset='${-1 * CHART_CIRCUMFERENCE * sliceOffset}'
+            stroke-width='${CHART_DIAMETER * (1 - innerRadius)}%'
+          ></circle>
+        `
+        acc.push(circle)
+        sliceOffset += percentage
+
+        return acc
+      }, init).join('')}
+    `
+  }
+
+  createChartLabels () {
+    return `
+      ${this.title !== ''
+        ? `<strong class='circle-chart-title'>${this.title}</strong>`
+        : ''}
+      ${this.data?.map(({ name, color }) => (
+        `<label class='circle-chart-label' style='--color: ${color}'>${name}</label>`
+      )).join('')}
+    `
+  }
+
+  createCircleLabel () {
+    const relevantItem = this.data?.find(d => d.key === this.showPercentFor)
+    if (!relevantItem) {
+      return ''
+    }
+    const percentage = calcPercentage(this.total, relevantItem.count)
+    return `
+      <text
+        dy='${CHART_RADIUS * 0.15}'
+        fill='${relevantItem.color}'
+        font-size='${CHART_RADIUS * 0.4}'
+        font-size='50'
+        x='${CHART_RADIUS}'
+        y='${CHART_RADIUS}'
+      >
+          ${Math.round(percentage * 100)}%
+      </text>
+    `
+  }
+
+  createChart () {
+    // Create SVG with circles
+    const circles = this.composeCircles()
+    this.svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg')
+    this.svg.setAttribute('viewBox', `0 0 ${CHART_DIAMETER} ${CHART_DIAMETER}`)
+    this.svg.innerHTML = `
+      ${circles}
+      ${this.createCircleLabel()}
+    `
+
+    // Create labels
+    this.labels = document.createElement('figcaption')
+    this.labels.innerHTML = this.createChartLabels()
+
+    // Add chart elements to DOM
+    this.chartElement?.append(this.svg)
+    this.chartElement?.append(this.labels)
```

Comment:
Re: `this.chartElement?.`,

Is it possible `chartElement` is missing/`undefined`? Or was that just to appease the linting gods?

Reply:
I would not expect it to be missing or `undefined` and it really is to make the linter happy.